### DISCUSSION TITLE
Improved handling of long running queries

### DISCUSSION THREAD

**Discussion by sunker** [grafana/grafana#45325] [Data Sources]
**Problem**
For many AWS plugins, the APIs that they're integrating with are async. They usually follow this pattern:
1. `StartQuery`
2. `CheckQueryStatus` Usually done on an interval or using a backoff. This action would yield statuses such as `running`, `partial` or `complete`.
3. `GetQueryResult` - Return prepared data once the query status is complete. 

For plugins such as Athena and Redshift, we handle all three steps above in one single http request. In other words, they're using the standard execution flow for Grafana which means they're getting support for alerting, query caching and recorded queries automatically. 

For services such as Athena and Redshift, it's common for queries to be slow. That means the http connection might be opened for a long time and that is not ideal. Hosted Grafana has an http connection timeout of 15 minutes and in AMG that limit is 1 minute, which is causing many of the queries to be terminated before they are completed. 

**What I want**
Plugins should be able to execute long running queries without keeping one http connection open throughout the whole query execution. This should not break alerting, query caching, recorded queries and other capabilities that come out of the box while using the standard query execution flow. I'd also like there to be one, idiomatic way of achieving this using the data source plugin api. The solution should also be compatible with an HA setup.

**Comment by marefr** [comment]
Just to get the discussions going;

One alternative is that the client (browser) needs to keep track of these things. As long as integrated services holds the state HA shouldn't be a problem. For alerting or recorded queries it will be a regular query that needs to execute synchronously. 

In general I don't see how Grafana would understand in the backend whether a query is long running or not. Are you thinking this to be some property on the query itself?

Maybe @grafana/grafana-edge-squad @grafana/observability-squad have some input?

**Reply by gabor** [reply]
> In general I don't see how Grafana would understand in the backend whether a query is long running or not. Are you thinking this to be some property on the query itself?

i understood this as .. the datasource-plugin would somehow say that they support the slow-query-mode-api, that Grafana can discover, and if it is so, every query going to that plugin would go through the slow-query-mode-api...

**Reply by gabor** [reply]
hmm.. maybe i misunderstood where the "long http connection" is happening, see my later replies :)

**Comment by gabor** [comment]
sorry to expand the discussion even more, but if we want to support queries that take 10minutes or more to finish, we should think about the user-interface too.

if i run a query that takes 10minutes to return results in a dashboard or in explore-mode, the current user-interface is not good for those. imagine that you are waiting, and after 8minutes you by-mistake close your browser-tab 😿 

generally for something so long running we would have to implement a mailbox-ish system, where you start the query, and you can close your laptop, do something else, come back to grafana, you get a notification your query is finished, you click on the notification, you see the results.

**Comment by gabor** [comment]
as i think more about it, the problem with the http-connection-timeout is happening between the browser and grafana-server, right?

so the solution would have to be a new API for browser<>grafana_server communcation .. in theory this could be done transparently, so that the javascript-datasource-code does not need to be aware of it.. maybe.. (the UX problem i mentioned above still applies)

**Comment by andresmgot** [comment]
@marefr @sunker @gabor I have created a design document for this: https://docs.google.com/document/d/14fTxDzkuA11uRzYXmmzYiJq8ZwQ0scauLEofpW5e_tc

And I have started working on a POC for the Hackathon, let me know what you think!

**Reply by andresmgot** [reply]
As a summary for the POC, I was able to create an asynchronous flow as described in the document linked above. In the GIF below, you can see how an asynchronous query is made in two steps, the first starts the query and the second returns the result (which is just `off`):

![Peek 2022-03-11 16-47](https://user-images.githubusercontent.com/4025665/157901722-767ef0cd-27dd-4aca-a506-30630adb43c9.gif)

So far, the code I've written temporarily lives at:

https://github.com/grafana/redshift-datasource/tree/pocAsync
https://github.com/grafana/sqlds/tree/pocAsync
https://github.com/grafana/grafana-aws-sdk/tree/pocAsync

**Reply by marefr** [reply]
Nice :+1: Looks promising :tada:

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
